xkafka/consumer.go

```diff
@@ -0,0 +1,154 @@
+package xkafka
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"github.com/sourcegraph/conc/stream"
+)
+
+// KafkaConsumer is the interface for kafka consumer.
+type KafkaConsumer interface {
+	GetMetadata(topic *string, allTopics bool, timeoutMs int) (*kafka.Metadata, error)
+	ReadMessage(timeout time.Duration) (*kafka.Message, error)
+	SubscribeTopics(topics []string, rebalanceCb kafka.RebalanceCb) error
+	Unsubscribe() error
+	Close() error
+}
+
+// ConsumerFunc is a function that returns a KafkaConsumer.
+type ConsumerFunc func(cfg *kafka.ConfigMap) (KafkaConsumer, error)
+
+func (cf ConsumerFunc) apply(o *options) { o.consumerFn = cf }
+
+// DefaultConsumerFunc is the default ConsumerFunc that initialises
+// a new confluent-kafka-go/kafka.Consumer.
+func DefaultConsumerFunc(cfg *kafka.ConfigMap) (KafkaConsumer, error) {
+	return kafka.NewConsumer(cfg)
+}
+
+// Consumer manages the consumption of messages from kafka topics
+// and the processing of those messages.
+type Consumer struct {
+	name        string
+	kafka       KafkaConsumer
+	middlewares []middleware
+	config      options
+}
+
+// NewConsumer creates a new Consumer instance.
+func NewConsumer(name string, opts ...Option) (*Consumer, error) {
+	cfg := defaultConsumerOptions()
+
+	for _, opt := range opts {
+		opt.apply(&cfg)
+	}
+
+	_ = cfg.configMap.SetKey("bootstrap.servers", strings.Join(cfg.brokers, ","))
+
+	consumer, err := cfg.consumerFn(&cfg.configMap)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Consumer{
+		name:   name,
+		config: cfg,
+		kafka:  consumer,
+	}, nil
+}
+
+// GetMetadata returns the metadata for the consumer.
+func (c *Consumer) GetMetadata() (*kafka.Metadata, error) {
+	return c.kafka.GetMetadata(nil, false, int(c.config.metadataTimeout.Milliseconds()))
+}
+
+// Use appends a MiddlewareFunc to the chain.
+// Middleware can be used to intercept or otherwise modify, process or skip messages.
+// They are executed in the order that they are applied to the Consumer.
+func (c *Consumer) Use(mwf ...MiddlewareFunc) {
+	for _, fn := range mwf {
+		c.middlewares = append(c.middlewares, fn)
+	}
+}
+
+// Start consumes from kafka and dispatches messages to handlers.
+// It blocks until the context is cancelled or an error occurs.
+// Errors are handled by the ErrorHandler if set, otherwise they stop the consumer
+// and are returned.
+func (c *Consumer) Start(ctx context.Context, handler Handler) error {
```

Comment:
Let's take handler as a required parameter while creating the `*Consumer`. It will make the Start function not depend on handler. This is use-full in cases where utilities like [gojekfarm/xrun](https://github.com/gojekfarm/xrun) only need the context to start and stop the component.

Reply:
Added a builder method `WithHandler` method